### ISSUE TITLE
Avoid syntax where child state is specified on the parent

### ISSUE THREAD

**Issue by LeaVerou** [shoelace-style/webawesome#238]
While there are exceptions, it’s *usually* an antipattern for local child state to live on the parent:
- it means logic is not localized and needs to be maintained in two places. This becomes even more difficult to maintain for long components like `<wa-page>`
- it's easy to not keep them in sync, e.g. remove or edit the child and forget the attribute on the parent

This is about local, child-specific state. Specifying common things on an ancestor to avoid specifying them on every individual descendant is a best practice not an antipattern! 

Hopefully this will become more clear with some examples of where WA breaks it:
- `disable-sticky` on `<wa-page>`: since there are *always* slotted elements for the slots we want to disable sticky on, it makes sense to specify this on them.
- `nav-state` on `<wa-page>`: similarly, since there is always a nav slotted in for this to make sense, it can be specified there
- `value` in `<wa-radio-group>`, `<wa-select>`: While we should definitely have a `value` prop, the established pattern is to specify what is selected/checked via an attribute on the option
- `active` in `<wa-tab-group>`

Edit to clarify: these kinds of summary props can be useful as JS getters/setters, since these concerns do not apply when reading or writing values via JS, and not having them means having to do tedious DOM wrangling. This is about the HTML API and the ergonomics of writing HTML.

**Comment by claviska** [comment]
> value in <wa-radio-group>, <wa-select>: While we should definitely have a value prop, the established pattern is to specify what is selected/checked via an attribute on the option

This is true, but many users have found it easier for the parent element (the "controller") to get and set the value. It's a common pattern in component libraries, particularly (but not limited to) React.

The platform's pattern here is, IMO, much worse in terms of DX.

**Comment by claviska** [comment]
> disable-sticky on `<wa-page>`: since there are always slotted elements for the slots we want to disable sticky on, it makes sense to specify this on them.
> nav-state on `<wa-page>`: similarly, since there is always a nav slotted in for this to make sense, it can be specified there

I’m pretty OK with this. I’m less OK with this, despite it being a React-like pattern.

> value in `<wa-radio-group>`, `<wa-select>`: While we should definitely have a value prop, the established pattern is to specify what is selected/checked via an attribute on the option

Just last night, I once again struggled with getting the selected option’s value from a plain `<select>` and it happens EVERY TIME. I think this is why the controller/item pattern emerged.

I think what we’re doing is objectively better than the platform, e.g. setting/getting value on the host element. The options in a select are child elements, sure, but the select is what determines which ones are selected based on interactions. Same with radio.

Aside from that, I’ll add a complaint about the complexity of having to adjust this in all affected components to align with an inferior API design that even a seasoned web developer can’t remember off the top of his head.

That one will be a hard sell, but since it doesn’t affect `<wa-page>` we can split it out and discuss further at a later time.

**Comment by LeaVerou** [comment]
> > value in `<wa-radio-group>`, `<wa-select>`: While we should definitely have a value prop, the established pattern is to specify what is selected/checked via an attribute on the option
> 
> Just last night, I once again struggled with getting the selected option’s value from a plain `<select>` and it happens EVERY TIME. I think this is why the controller/item pattern emerged.
> 
> I think what we’re doing is objectively better than the platform, e.g. setting/getting value on the host element. The options in a select are child elements, sure, but the select is what determines which ones are selected based on interactions. Same with radio.

I think there's a misunderstanding here around what I was suggesting. I was talking about the HTML syntax. The controller element should *absolutely* have a writeable JS prop that gets you the current selected value and allows you to set it, just like `<select>` does (`.value` works, but also `.selectedOptions` for multi-selects…).

**Comment by claviska** [comment]
To clarify, are you proposing the primary change for radio group + radio selection is to add an attribute on the selected radio item?

**Comment by LeaVerou** [comment]
> To clarify, are you proposing the primary change for radio group + radio selection is to add an attribute on the selected radio item?

Two things:
1. Add an attribute on the selected radio item (`checked` if we want to be consistent with `<input type=radio>`, `selected` if we want to be consistent with `<option>`, leaning towards the former)
2. Do not reflect `value` on `<wa-radio-group>`

So, when setting the value via HTML, it is done via the property on the radio (thus not having to keep the actual value in sync), but when setting the value via JS, we can just assign it to `radioGroup.value` (and read `radioGroup.value` to figure it out without needing to look at the individual radios).

**Comment by claviska** [comment]
> Add an attribute on the selected radio item (checked if we want to be consistent with <input type=radio>, selected if we want to be consistent with <option>, leaning towards the former)

If the user adds the `checked` attribute on a radio, does the radio group update its value? What happens if they remove the `checked` attribute from the radio?

More importantly, what does this look like for React users and other frameworkers (another new word!) that control the DOM? Right now, they just bind to `value`. Will these reflected attributes cause rendering issues in those libraries? I'm not sure, so it would require time to test it out — I'm just not sure it's time well spent right now. 😕 

@KonnorRogers didn't radio items reflect their checked value? Was reflection removed for FACE changes (similar to disabled) or am I misremembering?

**Comment by claviska** [comment]
Just merged and backported into WA, btw: https://github.com/shoelace-style/shoelace/issues/2298

**Comment by LeaVerou** [comment]
> > disable-sticky on `<wa-page>`: since there are always slotted elements for the slots we want to disable sticky on, it makes sense to specify this on them.
> > nav-state on `<wa-page>`: similarly, since there is always a nav slotted in for this to make sense, it can be specified there
> 
> I’m pretty OK with this.

In that case we need to decide what the new API should look like, and this also depends on #266 since these would now be attributes that can be specified on non-WA elements.

`disable-sticky` needs some thinking, though I think a valued attribute to control stickiness might be the way to go and allows room for more fine-grained customization in the future than on/off. 

But for `nav-state`, I wonder if there's actually some ARIA attribute we could recommend people add? Using a standard attribute if one is available circumvents having to make a decision on #266 AND means we don't have to update the light DOM element to add any ARIA because the author has done it for us, without having to add more syntax to express their intent.

> Just merged and backported into WA, btw: [shoelace-style/shoelace#2298](https://github.com/shoelace-style/shoelace/issues/2298)

Heh, thanks for the additional example 😁 Added it to the first post. Are there others?

**Comment by claviska** [comment]
If you have a proposal, freel free to attach it here. I'll let @KonnorRogers determine the amount of work it will require and if it's feasible to do by next week. (I'm happy to take a stab at it myself, but he built it and would know better what's easy/hard.)

> I wonder if there's actually some ARIA attribute we could recommend people add?

Aside — as a pattern, we try to avoid asking users to use `aria-` attributes because the components handle those themselves. It may seem silly or redundant that, on occasion, we use something like `label="foo"` and essentially add `aria-label="foo"` for them, but it gives us consistency (since not all aria attributes apply to the host) and control over how and when we work with them.

**Comment by LeaVerou** [comment]
> Aside — as a pattern, we try to avoid asking users to use `aria-` attributes because the components handle those themselves. It may seem silly or redundant that, on occasion, we use something like `label="foo"` and essentially add `aria-label="foo"` for them, but it gives us consistency (since not all aria attributes apply to the host) and control over how and when we work with them.

I think that makes sense for elements we control, but when dealing with elements we don't control, it may be better to ask authors to add these (when it does not duplicate intent) than to sprout them ourselves.

**Comment by KonnorRogers** [comment]
> > Add an attribute on the selected radio item (checked if we want to be consistent with , selected if we want to be consistent with , leaning towards the former)
> 
> If the user adds the `checked` attribute on a radio, does the radio group update its value? What happens if they remove the `checked` attribute from the radio?
> 
> More importantly, what does this look like for React users and other frameworkers (another new word!) that control the DOM? Right now, they just bind to `value`. Will these reflected attributes cause rendering issues in those libraries? I'm not sure, so it would require time to test it out — I'm just not sure it's time well spent right now. 😕
> 
> @KonnorRogers didn't radio items reflect their checked value? Was reflection removed for FACE changes (similar to disabled) or am I misremembering?

I don't recall if I got around to fixing radios defaultChecked, I'd have to look

**Comment by KonnorRogers** [comment]
> disable-sticky on <wa-page>: since there are always slotted elements for the slots we want to disable sticky on, it makes sense to specify this on them.

Its important to note `position: sticky;` is attached to wrapper elements inside of the elements shadow root, and are not directly applied to page children.

This is for a few reasons, the big ones being we need a ResizeObserver on sticky elements to know how much "top" space theyre taking up so we can set a CSS variable all the other sticky elements can "inherit" and set `top: Npx` so on page scroll they are the correct number of pixels from the top.

I do wonder if "sticky by default" is too presumptuous as it was mainly approached from a docs site / page with sidebar perspective. I'm not sure its the correct behavior, but getting sticky to work just right is definitely a challenge.

**Comment by LeaVerou** [comment]
> > disable-sticky on : since there are always slotted elements for the slots we want to disable sticky on, it makes sense to specify this on them.
> 
> Its important to note `position: sticky;` is attached to wrapper elements inside of the elements shadow root, and are not directly applied to page children.

Oh I see, so then we can't address them with regular CSS, right? (I imagine `:has(> ::slotted([foobar]))` doesn't work, right? Or does it?)

> I do wonder if "sticky by default" is too presumptuous as it was mainly approached from a docs site / page with sidebar perspective. I'm not sure its the correct behavior, but getting sticky to work just right is definitely a challenge.

I think it was the right call, though I'm less sure about whether subheader should have been one of these. But I'm not sure about subheader in general tbh…

**Comment by KonnorRogers** [comment]
> Oh I see, so then we can't address them with regular CSS, right? (I imagine :has(> ::slotted([foobar])) doesn't work, right? Or does it?)

The spec is changing all the time around this and my best guess would be at best its supported in some browsers, and at worst, not supported in any.

> I think it was the right call, though I'm less sure about whether subheader should have been one of these. But I'm not sure about subheader in general tbh…

Yea subheader is because i saw a number of docs sites that will have the sub header sticky, and then the header not sticky 🤔 

For example, here's the Vue docs on mobile viewports:


https://github.com/user-attachments/assets/866321ec-5276-4e57-a381-62bc5ade7362

**Comment by claviska** [comment]
Is there anything actionable that came from this? If so, can we break it into separate issues and close this one?

**Comment by LeaVerou** [comment]
@claviska I think the main one is the one about `value` and the one about `active`. The `<wa-page>` stuff needs more design work to be actionable and I'm unsure about the I/E of those.
I'll open new issues.

**Comment by LeaVerou** [comment]
Reopening this as it now has child issues, so it's weird to have the parent closed.
